protocol.md

```diff
@@ -0,0 +1,196 @@
+
+# Protocol Overview
+
+Notice: this protocol is documented post factum. The goal of it is to clearly present the current design and prepare the ground for its second version.
+
+The implementation of this protocol is mainly done in https://github.com/status-im/status-react and https://github.com/status-im/status-go repositories.
+
+The goal of this protocol is to allow people running Ethereum nodes with Whisper service enabled to exchange messages that are encrypted end-to-end in a way that guarantees darkness to some extent (messages are broadcasted through the network rather than send directly or using a known path).
+
+It's important to notice that messages are not limited to be text messages only. They can also have special meaning depending on the client implementation. For example, in the current implementation, there are message which informs about Eth requests or current user's location.
+
+This protocol consist of two components:
+* payload specification
+* Whisper adapter.
+
+The payload specification describes how the messages are encoded and decoded and what each fields mean. This is required to properly interpret messages by the client.
+
+Whisper adapter specifies interaction with the Whiper service with regards to keys management, configuration and some metadata (like topic) required to properly process and encrypt/decrypt messages.
+
+The protocol does not specify things like peers discovery, however, some notes and the current implementation will be described in an appendix.
+
+# Payload
+
+Payload is encrypted and decrypted using [transit format](https://github.com/cognitect/transit-format). It is a text-based format so that each decrypted message's payload can be easily read.
+
+Payload contains the following fields:
+* text `string`
+* content type `enum { text/plain }`
+* message type `enum { public-group-user-message, user-message }`
+* clock `int64`
+* timestamp `int64`
+* content `struct { chat-id string, text string }`
+
+Example of a valid message:
+
+```
+["~#c4",["abc123","text/plain","~:public-group-user-message",154593077368201,1545930773682,["^ ","~:chat-id","testing-adamb","~:text","abc123"]]]
+```
+
+As you can see, the message is an array and each index value has its meaning:
+* 0: `c4` is a decoder handler identificator for the current payload format. Identificators allow to register handler for many different types of messages
+* 1: array which items correspond to the described payload fields above
+
+For more details regarding serialization and deserialization please consult [transit format](https://github.com/cognitect/transit-format) specification.
+
+## Content types
+
+Here is a list of supported content types:
+* `text/plain` identifies a message which content is a plain text
+
+TODO: are there other types? How is it useful? Sending Eth requests?
```

Comment:
maybe @yenda or @rasom know?

Reply:
There is many other types:
```clojure
(def content-type-text "text/plain")
(def content-type-sticker "sticker")
(def content-type-status "status")
(def content-type-command "command")
(def content-type-command-request "command-request")
(def content-type-emoji "emoji")
```

Reply:
It is useful at the UI level to render messages differently based on their content type

Reply:
@yenda thanks! Can you please also provide some information what is each of these types?

Also @oskarth @mandrigin, the question here is whether we should support a strict list of content types or select a few as required ones and mention that other content types can be added by specific implementations?

Reply:
Hm sure, not sure best way to do this. It's a bit app specific, so kind of like extensions. I.e. stickers might not work on CLI.  Maybe phrase as MAY support and SHOULD textplain?

Reply:
We will address specific content types in the next PR. Some should be required but definitely not all of them.